### PR TITLE
[pkg/util/log] Fix a race condition at startup

### PR DESCRIPTION
**Pull request by L3n41c**
### What does this PR do?

Turn the `"pkg/util/log".Logger` variable from a raw pointer `*DatadogLogger` to an atomic pointer `atomic.Pointer[DatadogLogger]`.
Get rid of `"pkg/util/log".bufferLogsBeforeInit` because it contained exactly the same thing as `Logger == nil`.

### Motivation

Our [end-to-end tests are currently very flaky because the standalone `dogstatsd` process is regularly crashing](https://app.datadoghq.com/ci/test-runs?query=test_level%3Atest%20%40test.service%3Adatadog-agent%20%40git.branch%3Amain%20%40test.suite%3A%22github.com%2FDataDog%2Fdatadog-agent%2Ftest%2Fnew-e2e%2Ftests%2Fcontainers%22%20%28%40test.name%3A%22TestEKSSuite%2FTest00UpAndRunning%2Fagent_pods_are_ready_and_not_restarting%22%20OR%20%40test.name%3A%22TestKindSuite%2FTest00UpAndRunning%2Fagent_pods_are_ready_and_not_restarting%22%29&citest_explorer_sort=time%2Cdesc&cols=%40test.status%2Ctimestamp%2C%40test.suite%2C%40test.name%2C%40duration%2C%40test.service%2C%40git.branch&currentTab=overview&eventStack=AgAAAYxov5ggmJeQTgAAAAAAAAAYAAAAAEFZeG92NWdnQUFEN25IcjVtSWJDcTg4agAAACQAAAAAMDE4YzY4YzktYjlkZi00NDgwLThkNzQtMTMyMGZkODAzNThj&index=citest&mode=sliding&saved-view-id=2236559&testId=AgAAAYxov5ggmJeQTgAAAAAAAAAYAAAAAEFZeG92NWdnQUFEN25IcjVtSWJDcTg4agAAACQAAAAAMDE4YzY4YzktYjlkZi00NDgwLThkNzQtMTMyMGZkODAzNThj&trace=AgAAAYxov5ggmJeQTgAAAAAAAAAYAAAAAEFZeG92NWdnQUFEN25IcjVtSWJDcTg4agAAACQAAAAAMDE4YzY4YzktYjlkZi00NDgwLThkNzQtMTMyMGZkODAzNThj&start=1701969423665&end=1702574223665&paused=false) at startup.

Each time there is a crash, the `dogstatsd` logs are always showing a `fatal error: sync: Unlock of unlocked RWMutex` error in `pkg/util/log`.

Here are the full stack traces of such a fatal error that happened at [`pkg/util/log/log.go:481`](https://github.com/DataDog/datadog-agent/blob/fd0c71f19d990828bfc76b607c3c6e168b67829b/pkg/util/log/log.go#L473-L481):
* [ci-25049919-4670-kind-cluster](https://dddev.datadoghq.com/logs?query=kube_cluster_name%3Aci-25049919-4670-kind-cluster%20kube_namespace%3Adogstatsd-standalone%20&cols=host%2Cservice&index=%2A&messageDisplay=inline&refresh_mode=paused&stream_sort=desc&viz=stream&from_ts=1702494900000&to_ts=1702494960000&live=false)

Here are two other examples of such a fatal error that happened at [`pkg/util/log/log.go:450`](https://github.com/DataDog/datadog-agent/blob/fd0c71f19d990828bfc76b607c3c6e168b67829b/pkg/util/log/log.go#L441-L450):
* [ci-25045405-4670-kind-cluster](https://dddev.datadoghq.com/logs?query=kube_cluster_name%3Aci-25045405-4670-kind-cluster%20kube_namespace%3Adogstatsd-standalone%20&cols=host%2Cservice&index=%2A&messageDisplay=inline&refresh_mode=paused&stream_sort=desc&viz=stream&from_ts=1702491180000&to_ts=1702491240000&live=false)
* [ci-25038804-4670-kind-cluster](https://dddev.datadoghq.com/logs?query=kube_cluster_name%3Aci-25038804-4670-kind-cluster%20kube_namespace%3Adogstatsd-standalone%20&cols=host%2Cservice&index=%2A&messageDisplay=inline&refresh_mode=paused&stream_sort=desc&viz=stream&from_ts=1702487760000&to_ts=1702487820000&live=false)

```
fatal error: sync: Unlock of unlocked RWMutex

goroutine 232 [running]:
sync.fatal({0x1adf24f?, 0x953f25?})
    /usr/local/go/src/runtime/panic.go:1031 +0x1e
sync.(*RWMutex).Unlock(0xc000690d20)
    /usr/local/go/src/sync/rwmutex.go:209 +0x4a
github.com/DataDog/datadog-agent/pkg/util/log.logFormat(0x0?, 0x0?, 0xc000959e98, {0x1aeb60b, 0x25}, {0x0, 0x0, 0x0})
    /go/src/github.com/DataDog/datadog-agent/pkg/util/log/log.go:481 +0x16d
github.com/DataDog/datadog-agent/pkg/util/log.Infof({0x1aeb60b, 0x25}, {0x0, 0x0, 0x0})
    /go/src/github.com/DataDog/datadog-agent/pkg/util/log/log.go:626 +0xee
```

By looking at the code, it seems that the `Logger` global variable can be concurrently written by the `SetupLogger` function here: https://github.com/DataDog/datadog-agent/blob/fb0bad2f98964e5f2c47cd975844b3d147466b83/pkg/util/log/log.go#L56-L57
and be read by the `Log` function here: https://github.com/DataDog/datadog-agent/blob/fb0bad2f98964e5f2c47cd975844b3d147466b83/pkg/util/log/log.go#L441-L442
or by the `logFormat` function here: https://github.com/DataDog/datadog-agent/blob/fb0bad2f98964e5f2c47cd975844b3d147466b83/pkg/util/log/log.go#L473-L474

The evidences that the crashes come from a race with `SetupLogger` are the following:
In the `dogstatsd` logs, we can see:
```
2023-12-13 17:15:31 UTC | DSD | INFO | (comp/core/workloadmeta/store.go:498 in startCandidates) | workloadmeta collector "podman" could not start. error: component workloadmeta-podman is disabled: Podman not detected
2023-12-13 17:15:25 UTC | DSD | INFO | (pkg/util/log/log.go:626 in func1) | 3 Features detected from environment: kubernetes,cri,containerd
2023-12-13 17:15:25 UTC | DSD | INFO | (pkg/util/kubernetes/kubelet/kubelet_client.go:289 in checkKubeletConnection) | Successful configuration found for Kubelet, using URL: https://172.18.0.2:10250
```

Whereas the file, line number and function are correct for the first and third logs, the second log coming from [`pkg/config/env/environment_detection.g:L124`](https://github.com/DataDog/datadog-agent/blob/4427046328c2239274e7300e51dba28ddf773996/pkg/config/env/environment_detection.go#L124) is said to come from [`pkg/util/log/log.go:626`](https://github.com/DataDog/datadog-agent/blob/4427046328c2239274e7300e51dba28ddf773996/pkg/util/log/log.go#L626) but from an anonymous function `func1` !
This is the sign that there’s an extra frame in the call stack due to the queuing on an “early” log inside `logsBuffer`.

And what is super interesting is that this `… Features detected from environment …` log is supposed to be emitted only once at the `dogstatsd` startup.

And if we look at all the [`dogstatsd` logs here](https://dddev.datadoghq.com/logs?query=kube_cluster_name%3Aci-%2A%20kube_namespace%3Adogstatsd-standalone%20%28%22Features%20detected%20from%20environment%22%20OR%20%22Unlock%20of%20unlocked%20RWMutex%22%29%20&cols=host%2Cservice&index=%2A&messageDisplay=inline&refresh_mode=sliding&stream_sort=time%2Cdesc&viz=stream&from_ts=1702387835670&to_ts=1702560635670&live=true), we can see that there’s a perfect match between:
* when the `… Features detected from environment …` log is logged twice instead of once (check the host and timestamp) and
* when there is a `fatal error: sync: Unlock of unlocked RWMutex` error.

### Additional Notes

In this PR, I didn’t try to be smart and revisit the locking strategy. I only mechanically replaced the raw pointer `*DatadogLogger` by an atomic pointer `atomic.Pointer[DatadogLogger]` for the `Logger` variable because its accesses were not guarded by any mutex.

### Describe how to test/QA your changes

I didn’t manage to have a reproducer that triggers the error as often as in our end-to-end tests, but I still managed to reproduce the error in a relatively short period of time with this:

```shell
truncate -s 0 /tmp/dogstatsd.log; while true; do echo -n .; DD_LOG_FILE=/dev/null DD_URL=http://localhost:8042 timeout 1 bin/dogstatsd/dogstatsd start > /tmp/dogstatsd.log 2>&1; grep 'Unlock of unlocked RWMutex' /tmp/dogstatsd.log; done
```

It starts `dogstatsd` in a loop.
Each time it starts fine, it prints a dot `.`.
And when there is an error it also prints `Unlock of unlocked RWMutex`.

Here is what I got with a `dogstatsd` built from `main`:
image
We can see that I managed to reproduce the error quite a few times.

And here is what I got with this PR:
image
I haven’t been able to see any error anymore.

### Reviewer's Checklist

- [ ] If known, an appropriate milestone has been selected; otherwise the `Triage` milestone is set.
- [ ] Use the `major_change` label if your change either has a major impact on the code base, is impacting multiple teams or is changing important well-established internals of the Agent. This label will be use during QA to make sure each team pay extra attention to the changed behavior. For any customer facing change use a releasenote.
- [ ] A [release note](https://github.com/DataDog/datadog-agent/blob/main/docs/dev/contributing.md#reno) has been added or the `changelog/no-changelog` label has been applied.
- [ ] Changed code has automated tests for its functionality.
- [ ] Adequate QA/testing plan information is provided if the `qa/skip-qa` label is not applied.
- [ ] At least one `team/..` label has been applied, indicating the team(s) that should QA this change.
- [ ] If applicable, docs team has been notified or [an issue has been opened on the documentation repo](https://github.com/DataDog/documentation/issues/new).
- [ ] If applicable, the `need-change/operator` and `need-change/helm` labels have been applied.
- [ ] If applicable, the `k8s/<min-version>` label, indicating the lowest Kubernetes version compatible with this feature.
- [ ] If applicable, the [config template](https://github.com/DataDog/datadog-agent/blob/main/pkg/config/config_template.yaml) has been updated.

### CODE REVIEW

**Review by dustmop — APPROVED** [review]
LGTM

**Review by dustmop — CHANGES_REQUESTED** [review]
Actually, nevermind, I saw [this linter](https://app.circleci.com/pipelines/github/DataDog/datadog-agent/109781/workflows/771aec46-4492-4f52-8b5e-390bf60f11ec/jobs/1313975) is suggesting a different atomic library.

```
Linter failures:
pkg/util/log/log.go:24:2: import 'sync/atomic' is not allowed from list 'main': Use go.uber.org/atomic instead; see docs/dev/atomics.md (depguard)
        "sync/atomic"
```

Could you try using this instead and running your test again?

**Comment by pgimalac** [comment]
I know someone asked about removing that lint to enable that lib again, but it doesn't look like they did

**Comment by L3n41c** [comment]
@dustmop I’ve switched to `go.uber.org/atomic` as a replacement for `sync/atomic` in https://github.com/DataDog/datadog-agent/pull/21573/commits/f552c1ce69dc636eb2a70703b00e2bc83495be03.

**Review by dustmop — APPROVED** [review]
Please also run `inv tidy-all` to fix the remaining linter failures. After that, LGTM

**Review by sblumenthal — APPROVED** [review]

File: pkg/util/log/log.go [hunk 1]
```diff
@@ -120,7 +119,7 @@ func (sw *DatadogLogger) changeLogLevel(level string) error {
 	if !ok {
 		return errors.New("bad log level")
 	}
-	Logger.level = lvl
+	logger.Load().level = lvl
```
Comment (zhuminyi) [inline pkg/util/log/log.go:122]: Do you need to check nil here?

Comment (L3n41c) [reply pkg/util/log/log.go:122]: The change of the type of pointer of the `logger` variable didn’t change the execution flow.
So, if checking `nil` wasn’t needed before, it shouldn’t be needed now either.
The code must probably already guarantee that `changeLogLevel` cannot be called before `SetupLogger`.

**Review by vickenty — APPROVED** [review]

**Review by mfpierre — APPROVED** [review]

**Review by coignetp — APPROVED** [review]

**Review by chouetz — APPROVED** [review]

**Pull request merged**